tracer/src/Datadog.Trace/Agent/StatsBuffer.cs

```diff
@@ -0,0 +1,178 @@
+using System.IO;
+using Datadog.Trace.ExtensionMethods;
+using Datadog.Trace.Vendors.Datadog.Sketches;
+using Datadog.Trace.Vendors.MessagePack;
+
+namespace Datadog.Trace.Agent
+{
+    internal class StatsBuffer
+    {
+        private readonly List<StatsAggregationKey> _keysToRemove;
+
+        private readonly ClientStatsPayload _header;
+
+        public StatsBuffer(ClientStatsPayload header)
+        {
+            _header = header;
+            _keysToRemove = new();
+            Buckets = new();
+            Reset();
+        }
+
+        public Dictionary<StatsAggregationKey, StatsBucket> Buckets { get; }
+
+        public long Start { get; private set; }
+
+        public void Reset()
+        {
+            // We need to do some cleanup because the application could have an unlimited number of endpoints,
+            // but at the same time we don't want to reallocate all the sketches every time.
+            // The compromise here is to remove only the endpoints that received no hit during the last iteration.
+            foreach (var kvp in Buckets)
+            {
+                if (kvp.Value.Hits == 0)
+                {
+                    _keysToRemove.Add(kvp.Key);
+                }
+                else
+                {
+                    kvp.Value.Clear();
+                }
+            }
+
+            foreach (var key in _keysToRemove)
+            {
+                Buckets.Remove(key);
+            }
+
+            _keysToRemove.Clear();
+
+            Start = DateTimeOffset.UtcNow.ToUnixTimeNanoseconds();
+        }
+
+        public void Serialize(Stream stream, long duration)
+        {
+            MessagePackBinary.WriteMapHeader(stream, 9);
+
+            MessagePackBinary.WriteString(stream, "Hostname");
+            MessagePackBinary.WriteString(stream, _header.HostName ?? string.Empty);
+
+            MessagePackBinary.WriteString(stream, "Env");
+            MessagePackBinary.WriteString(stream, _header.Environment ?? string.Empty);
+
+            MessagePackBinary.WriteString(stream, "Version");
+            MessagePackBinary.WriteString(stream, _header.Version ?? string.Empty);
+
+            MessagePackBinary.WriteString(stream, "Stats");
+            MessagePackBinary.WriteArrayHeader(stream, 1);
+            SerializeBuckets(stream, duration);
+
+            MessagePackBinary.WriteString(stream, "Lang");
+            MessagePackBinary.WriteString(stream, TracerConstants.Language);
+
+            MessagePackBinary.WriteString(stream, "TracerVersion");
+            MessagePackBinary.WriteString(stream, TracerConstants.AssemblyVersion);
+
+            MessagePackBinary.WriteString(stream, "RuntimeID");
+            MessagePackBinary.WriteString(stream, Tracer.RuntimeId);
+
+            MessagePackBinary.WriteString(stream, "Sequence");
+            MessagePackBinary.WriteInt64(stream, _header.GetSequenceNumber());
+
+            MessagePackBinary.WriteString(stream, "Service");
```

Comment:
We're missing [the `"agentAggregation"` key](https://github.com/DataDog/datadog-agent/blob/5ac8dd026cd315fbf383ec9c72cfcac6338d708b/pkg/trace/pb/stats.proto#L33), is that ok?

Reply:
Yeah, this one is only set by the agent (and missing == empty in messagepack)